Data object migration: Value migrations

LukasHuser opened the pull request:
New interface IDoValueMigrationHandler to support data object value migrations. In contrast to IDoStructureMigrationHandler, value migration handlers must not change the structure of a data object, but can replace values of specific attributes.

A value migration handler operates on typed values. Values can be of any type, typical use cases include merges, replacements or renaming of built-in IDs, such as IUuIds. However, values can also be IDoEntity data objects.

Value migration handlers are always applied after all structure migration handlers have been applied. Value migrations are applied in a specific order, defined by a type version. Migration handlers with the same type version are applied in arbitrary, but deterministic order.

315763

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationInventory.java:
```diff
@@ -77,6 +79,11 @@ public class DoStructureMigrationInventory {
   protected final Map<String, NamespaceVersion> m_typeNameToCurrentTypeVersion = new HashMap<>();
 
+  /**
+   * Ordered set of all value migration handlers.
```
Ordered map

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationInventory.java:
```diff
@@ -77,6 +79,11 @@ public class DoStructureMigrationInventory {
+  /**
+   * Ordered set of all value migration handlers.
+   */
+  protected final LinkedHashMap<DoValueMigrationId, IDoValueMigrationHandler<?>> m_valueMigrationHandlers = new LinkedHashMap<>();
```
Why not consistent with m_structureMigrationHandlers and use a Map from NamespaceVersion first?

LukasHuser replied:
Lookup by NamespaceVersion is not needed for value migration handlers. We have two requirements (which both can be met by using a LinkedHashMap):
1) Ordered list of all value migration handlers
2) Lookup of value migration handers by ID

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationInventory.java:
```diff
@@ -199,10 +213,17 @@ protected void init() {
     }
+
+    // Collect value migration handlers
+    getAllValueMigrationHandlers().stream()
```
Validation is missing here, e.g. unique value migration ID, known namespace version (especially important because comparator will fail for unknown ones)

LukasHuser replied:
Validation added

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationInventory.java:
```diff
@@ -107,12 +114,19 @@ protected Collection<Class<? extends IDoStructureMigrationTargetContextData>> ge
+  /**
+   * @return All value migration handlers.
+   */
+  protected List<IDoValueMigrationHandler> getAllValueMigrationHandlers() {
```
IDoValueMigrationHandler<?>

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationInventory.java:
```diff
@@ -199,10 +213,17 @@ protected void init() {
+    // Collect value migration handlers
+    getAllValueMigrationHandlers().stream()
+        .sorted(Comparator
+            .comparing(handler -> BEANS.get(((IDoValueMigrationHandler<?>) handler).typeVersion()).getVersion(), m_comparator) // primary sort order: type version
```
why BEANS.get? handler is an instance, not a class

LukasHuser replied:
BEANS.get was called on the Type Version. Changed implementation of IDoValueMigrationHandler and AbstractDoValueMigrationHandler to match existing pattern in IDoStructureMigrationHandler and AbstractDoStructureMigrationHandler: typeVersion() method returns a NamespaceVersion instead of Class<? extends ITypeVersion>.

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationInventory.java:
```diff
@@ -304,6 +325,23 @@ protected void validateDataObjectTypeVersion(IDoStructureMigrationHandler migrat
+   * @return ordered list of all available {@link IDoValueMigrationHandler}s, except for
+   *         {@code ignoredValueMigrationIds} which are filtered from the returned list
+   */
+  public List<IDoValueMigrationHandler<?>> getValueMigrationHandlers(Set<DoValueMigrationId> ignoredValueMigrationIds) {
```
I'd place the ignore handling in a helper or somewhere else. In my opinion, this shouldn't be part of the inventory itself.

LukasHuser replied:
Filtering of ignored handlers moved to callers

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationInventory.java:
```diff
@@ -304,6 +325,23 @@ protected void validateDataObjectTypeVersion(IDoStructureMigrationHandler migrat
+  /**
+   * @return {@link IDoValueMigrationHandler} for a given {@link DoValueMigrationId}
+   */
+  public IDoValueMigrationHandler<?> getValueMigrationHandler(DoValueMigrationId valueMigrationId) {
```
Check callers, where is this method use?

LukasHuser replied:
This method allows lookup of value migration handlers by ID and is useful for downstream projects when implementing a LookupCall for a SmartField or ListBox

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrator.java:
```diff
@@ -63,52 +64,46 @@ public <T extends IDataObject> DoStructureMigratorResult<T> migrateDataObject(Do
    *          Raw data object, might be a partial non-raw data object (i.e. _type info on certain entities). Only raw
    *          data object parts are migrated.
+   * @param valueType
+   *          The migrated object is cast to {@code valueType}. Can be {@link IDataObject} or a more specific sub-class.
```
"The migrated object is read into {@code valueType}." Cast might be misleading, because the data object given is untyped. Deserialization via Jackson takes place here.

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrator.java:
```diff
@@ -123,41 +118,73 @@ public boolean migrateDataObject(DoStructureMigrationContext ctx, IDataObject da
-      logger.debug("No data object entities with a type name found within {}", dataObject);
-      stats.addMigrationDuration(start);
-      return false;
+      ctx.getLogger().debug("No data object entities with a type name found within {}", dataObject);
```
return is missing here

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrator.java:
```diff
@@ -123,41 +118,73 @@ public boolean migrateDataObject(DoStructureMigrationContext ctx, IDataObject da
     }
-
     List<NamespaceVersion> versions = BEANS.get(DoStructureMigrationInventory.class).getVersions(typeVersions, toVersion);
-    if (versions.isEmpty()) {
```
Keep the return here, no need to iterate over an empty list and the if below for log debug can be omitted too

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrator.java:
```diff
@@ -123,41 +118,73 @@ public boolean migrateDataObject(DoStructureMigrationContext ctx, IDataObject da
+    boolean structureChanged = false;
     for (NamespaceVersion version : versions) {
-      changed |= migrateDataObject(ctxCopy, version, dataObject);
+      MigrationDataObjectVisitor visitor = createStructureMigrationVisitor(ctx, version);
```
why not keeping the protected method?

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrator.java:
```diff
@@ -123,41 +118,73 @@ public boolean migrateDataObject(DoStructureMigrationContext ctx, IDataObject da
+    return new DoStructureMigratorResult<>(migratedDataObject, objectChanged);
+  }
+
+  protected boolean applyStructureMigration(DoStructureMigrationContext ctx, IDataObject dataObject, NamespaceVersion toVersion) {
```
plural? applyStructureMigrations, for applyValueMigrations too? not sure though

LukasHuser replied:
Decision: we keep the singular names.

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrator.java:
```diff
@@ -123,41 +118,73 @@ public boolean migrateDataObject(DoStructureMigrationContext ctx, IDataObject da
+    return new DoStructureMigratorResult<>(migratedDataObject, valueMigrationVisitor.isChanged());
   }
 
-  protected boolean migrateDataObject(DoStructureMigrationContext ctx, NamespaceVersion version, IDataObject dataObject) {
```
Keep this method, see comment above. Or maybe a conflict with the deprecated method below?

LukasHuser replied:
All migrateDataObject methods returning a boolean value are moved to the bottom and have been deprecated. The existing protected method has been renamed to applyStructureMigration (consistent with the new method applyValueMigration)

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrator.java:
```diff
@@ -176,4 +203,40 @@ public boolean isChanged() {
+
+  /**
+   * This method only applies structure migrations. If this is the desired behavior, e.g. for unit tests, use
+   * {@link #applyStructureMigration(DoStructureMigrationContext, IDataObject, NamespaceVersion)} Or call one of the
```
Or -> or

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoValueMigrationDataObjectVisitor.java:
```diff
@@ -0,0 +1,72 @@
+
+    DoValueMigrationIdsContextData contextData = m_ctx.getGlobal(DoValueMigrationIdsContextData.class);
+    assertNotNull(contextData.getAppliedValueMigrationIds(), "Applied value migration IDs on context required.");
+    m_valueMigrationHandlers = m_inventory.getValueMigrationHandlers(contextData.getAppliedValueMigrationIds());
```
Handle exclusions here, not via inventory, see comment in inventory

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoValueMigrationDataObjectVisitor.java:
```diff
@@ -0,0 +1,72 @@
+      if (handler.valueClass().isAssignableFrom(currentValue.getClass())) {
+        // noinspection unchecked
+        T migratedValue = ((IDoValueMigrationHandler<T>) handler).migrate(m_ctx, currentValue);
+        m_changed |= ObjectUtility.notEquals(currentValue, migratedValue);
```
equals most likely doesn't work if migration handler changes value within directly, e.g. when acting on a data object. Change signature of migrate? To be discussed. Enforced via JavaDoc on migration handler, might be okay, still to be discussed.

LukasHuser replied:
For simple ID types such as StringId, the current signature is very convenient. When acting on data objects, the implementation must make sure to return a new instance or a modified copy of the provided input value. JavaDoc and code examples on IDoValueMigrationHandler provided for clarity

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoValueMigrationId.java:
```diff
@@ -0,0 +1,47 @@
+
+  private static final long serialVersionUID = 1L;
+
+  public static DoValueMigrationId create() {
```
Remove, should never be called, only used for fixed IDs

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoValueMigrationIdsContextData.java:
```diff
@@ -0,0 +1,36 @@
+  // thead-safe because read-only after initialization
+  protected Set<DoValueMigrationId> m_appliedValueMigrationIds;
+
+  public DoValueMigrationIdsContextData withAppliedValueMigrationIds(Set<DoValueMigrationId> valueMigrationIds) {
```
rename parameter to appliedValueMigrationIds

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoValueMigrationIdsContextData.java:
```diff
@@ -0,0 +1,36 @@
+    return this;
+  }
+
+  public Set<DoValueMigrationId> getAppliedValueMigrationIds() {
```
JavaDoc: @return Unmodifiable set of applied value migration IDs

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoStructureMigrationTargetContextData.java:
```diff
@@ -19,6 +19,7 @@
  * {@link DoStructureMigrationContextDataTarget} annotation for data object matching).
  */
 @Bean
+// TODO 23.1 [data object migration] rename to IDataObjectMigrationTargetContextData
```
This one shouldn't be renamed, only works for structure migration.

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoValueMigrationHandler.java:
```diff
@@ -0,0 +1,108 @@
+  DoValueMigrationId id();
+
+  /**
+   * @return Class of value node that needs to be migrated (Assignable check is performed so ExampleId will also be
```
(Assignable -> (assignable

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoValueMigrationHandler.java:
```diff
@@ -0,0 +1,108 @@
+public interface IDoValueMigrationHandler<T> {
+
+  /**
+   * Unique identifier for this migration handler.
```
for this value migration handler

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoValueMigrationHandler.java:
```diff
@@ -0,0 +1,108 @@
+   * M1 to make sure that M1 is applied <i>before</i> M2.
+   * <p>
+   * Example 2: Dependent value migrations in different namespaces<br>
+   * Value migration M1 from A -> B with type version crm-22.0.0<br>
```
crm -> lorem, project -> ipsum (or something else)

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoValueMigrationHandler.java:
```diff
@@ -0,0 +1,108 @@
+  /**
+   * Returns the migrated value for a given input value.
+   * <p>
+   * Clients of this method can detect changes to the migrated values by an equality check on the provided input value
```
Clients? Callers of this method?

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoValueMigrationHandler.java:
```diff
@@ -0,0 +1,108 @@
+ * <p>
+ * Value migration handlers are always applied <i>after</i> all {@link IDoStructureMigrationHandler} have been applied.
+ * Value migrations are applied in a specific order, defined by the given {@link #typeVersion()}. Migration handlers
+ * with the same type version are applied in arbitrary, but deterministic order.
```
can be influenced via @Order annotation?

LukasHuser replied:
Secondary sort order: Changed implementation from explicit order by simple class name to rely on order provided by the bean manager (ordered by @Order annotation and fully qualfied class name)

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoValueMigrationHandler.java:
```diff
@@ -0,0 +1,108 @@
+ *
+ *   &#64;Override
+ *   public Class<? extends ITypeVersion> typeVersion() {
+ *     return FooTypeVersions.Foo_2.class;
```
Remove FooTypeVersions here, static import usually

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/MigrationDataObjectVisitor.java:
```diff
@@ -24,6 +24,7 @@
  * Data object visitor to migrate data objects.
  */
+// TODO 23.1 [data object migration] rename to DoStructureMigrationDataObjectVisitor
 public class MigrationDataObjectVisitor extends AbstractDataObjectVisitor {
```
Rename already? Internal class, shouldn't be accessed from outside (protected?). To be discussed

LukasHuser replied:
Class renamed

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationContextDataTarget.java:
```diff
@@ -31,6 +31,7 @@
 @Retention(RUNTIME)
 @Target(TYPE)
+// TODO 23.1 [data object migration] rename to DataObjectMigrationContextDataTarget
 public @interface DoStructureMigrationContextDataTarget {
```
don't rename, only works for structure migration

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/fixture/house/RoomTypeValueMigrationHandler.java:
```diff
@@ -0,0 +1,45 @@
+
+  @Override
+  public Class<? extends ITypeVersion> typeVersion() {
+    return CharlieFixtureTypeVersions.CharlieFixture_2.class;
```
static import

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoValueMigrationHandler.java:
```diff
@@ -0,0 +1,108 @@
+ * exact point in time, when the structure migration was introduced.
+ * <p>
+ * Example: built-in ID FOO is merged / replaced by built-in ID BAR<br>
+ * Note: this migration is only idempotent, if value FOO is not used anymore in the future code base.
```
Add a note that, in constrast to structure migration handlers, a value migration should or must reference existing code, so that in case this code changes, the migration handler is changed too?

LukasHuser replied:
JavaDoc enhanced

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/fixture/house/RoomSizeValueMigrationHandler.java:
```diff
@@ -0,0 +1,54 @@
+
+  @Override
+  public Class<? extends ITypeVersion> typeVersion() {
+    return DeltaFixtureTypeVersions.DeltaFixture_2.class; // after RoomTypeValueMigrationHandler
```
static import

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/fixture/house/HouseValueMigrationHandler.java:
```diff
@@ -0,0 +1,58 @@
+
+  @Override
+  public Class<? extends ITypeVersion> typeVersion() {
+    return CharlieFixtureTypeVersions.CharlieFixture_1.class;
```
static import

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/fixture/house/HouseValueMigrationHandler.java:
```diff
@@ -0,0 +1,58 @@
+
+  @Override
+  public HouseFixtureDo migrate(DoStructureMigrationContext ctx, HouseFixtureDo value) {
+    RoomFixtureDo room = CollectionUtility.getElement(value.getRooms(), 0);
```
CollectionUtility.firstElement

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/fixture/house/HouseValueMigrationHandler.java:
```diff
@@ -0,0 +1,58 @@
+  @Override
+  public HouseFixtureDo migrate(DoStructureMigrationContext ctx, HouseFixtureDo value) {
+    RoomFixtureDo room = CollectionUtility.getElement(value.getRooms(), 0);
+    if (room != null && "tiny room".equals(room.getName())) {
```
Use early return here

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/fixture/house/HouseTypesFixture.java:
```diff
@@ -0,0 +1,24 @@
+ */
+package org.eclipse.scout.rt.dataobject.migration.fixture.house;
+
+public final class HouseTypesFixture {
```
Two classes, one for house types, one for room types. Or rename this class to HouseRoomTypesFixture or something similar? Probably two classes are the better approach here

LukasHuser replied:
Split into two classes

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/fixture/house/HouseTypeValueMigrationHandler.java:
```diff
@@ -0,0 +1,45 @@
+
+  @Override
+  public Class<? extends ITypeVersion> typeVersion() {
+    return CharlieFixtureTypeVersions.CharlieFixture_2.class;
```
static import

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoValueMigrationId.java:
```diff
@@ -0,0 +1,47 @@
+    return new DoValueMigrationId(UUID.randomUUID());
+  }
+
+  public static DoValueMigrationId of(UUID id) {
```
remove, not called

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoValueMigrationHandler.java:
```diff
@@ -0,0 +1,108 @@
+ * migration is applied multiple times. In general, value migrations are applied <i>exactly once</i>, as the system
+ * keeps track of all value migrations which have been applied already and won't be applied again. Idempotency is
+ * required, because structure migrations, which are applied before any value migration, might insert or change values
+ * as well - and those values might represent the source or the target of a given value migration, depending on the
```
Add note what possible options are available for non-idempotent migrations

LukasHuser replied:
JavaDoc enhanced.

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationStatsContextData.java:
```diff
@@ -99,7 +99,7 @@ public long getAccumulatedMigrationDurationNano() {
    *          Name to print for entities
    * @param entityCount
    *          Number of entities processed (optional), can be different than the number of calls made to
-   *          {@link DoStructureMigrator#migrateDataObject(DoStructureMigrationContext, IDataObject)}.
```
TODO with rename, check what other classes should be renamed too

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationTestHelper.java:
```diff
@@ -45,14 +48,15 @@
 
   public List<INamespace> getFixtureNamespaces() {
-    return Arrays.asList(new AlfaFixtureNamespace(), new BravoFixtureNamespace(), new CharlieFixtureNamespace());
+    return Arrays.asList(new AlfaFixtureNamespace(), new BravoFixtureNamespace(), new CharlieFixtureNamespace(), new DeltaFixtureNamespace());
```
if not required within this test, do not add those here, for type versions too

LukasHuser replied:
This helper class is used by the two test classes DataObjectMigratorValueMigrationTest and DataObjectMigratorStructureMigrationTest. At least DataObjectMigratorValueMigrationTest uses the DeltaFixtureNamespace and TypeVersions

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationContext.java:
```diff
@@ -70,12 +71,18 @@ protected DoStructureMigrationContext copy() {
    */
   protected DoStructureMigrationContext initializedCopy(IDoStructureMigrationLocalContextData... initialLocalContextDatas) {
-    DoStructureMigrationContext copy = copy();
+    return copy().withInitialLocalContext(initialLocalContextDatas);
```
remove the initializedCopy method completely. Add a JavaDoc to withInitialLocalContext that this method must only be called on a newly created context or on a copy {@link #copy}.

LukasHuser replied:
Removed method initializedCopy

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigratorTest.java:
```diff
@@ -68,21 +81,29 @@ public static void beforeClass() {
-        new RoomFixtureDoStructureMigrationHandler_5(),
-        new PersonFixtureDoStructureMigrationHandler_2());
+        Arrays.asList(new PostalAddressFixtureUpdateVersionOnlyMigrationHandler_2(),
+            new PetFixtureCaseSensitiveNameMigrationHandler_2(),
```
Split in two test classes, one for structure migrations, and for value migrations?
DataObjectMigratorStructureMigrationTest
DataObjectMigratorValueMigrationTest

LukasHuser replied:
Split into two separate test classes: DataObjectMigratorStructureMigrationTest and DataObjectMigratorValueMigrationTest

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigratorTest.java:
```diff
@@ -377,4 +399,138 @@ public void testListManipulationViaMigrationHandler() {
+        .put("areaInSquareMeter", 10)
+        .build();
+
+    DoStructureMigrationContext context = BEANS.get(DoStructureMigrationContext.class)
```
context -> ctx

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigratorTest.java:
```diff
@@ -377,4 +399,138 @@ public void testListManipulationViaMigrationHandler() {
+
+  /**
+   * Tests non-idempotent value migration with {@link RoomSizeValueMigrationHandler}.<br>
+   * Migration handler matches and replaces whole data object (RoomFixtureDo)
```
(RoomFixtureDo).

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigratorTest.java:
```diff
@@ -377,4 +399,138 @@ public void testListManipulationViaMigrationHandler() {
+
+    IDoEntity expected = BEANS.get(HouseFixtureDo.class)
+        .withName("example")
+        .withHouseType(HouseTypeFixtureStringId.of("detached-house"))
```
reference constant here

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigratorTest.java:
```diff
@@ -377,4 +399,138 @@ public void testListManipulationViaMigrationHandler() {
+    assertEqualsWithComparisonFailure(expected, result.getDataObject());
+
+    // migration run 2: data object is changed again, due to non-idempotent value migration
+    result = s_migrator.migrateDataObject(context, result.getDataObject(), RoomFixtureDo.class);
```
called with already typed object here, so no structure migrations will be applied, but should be okay for test

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigratorTest.java:
```diff
@@ -377,4 +399,138 @@ public void testListManipulationViaMigrationHandler() {
+   * Migration handler matches and replaces whole data object (RoomFixtureDo)
+   */
+  @Test
+  public void testReplaceDataObject() {
```
Combine a test that adds a value via structure migration that is migrated via value migration

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DataObjectMigratorStructureMigrationTest.java:
```diff
@@ -46,15 +49,17 @@
 /**
- * Tests for {@link DoStructureMigrator}.
+ * Tests for {@link DoStructureMigrator}, with focus on data object structure migrations
+ * ({@link IDoStructureMigrationHandler})
```
Add .

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DataObjectMigratorStructureMigrationTest.java:
```diff
@@ -68,21 +73,25 @@ public static void beforeClass() {
 
-    s_migrationContext = BEANS.get(DoStructureMigrationContext.class);
+    s_migrationContext = BEANS.get(DoStructureMigrationContext.class)
+        .putGlobal(BEANS.get(DoValueMigrationIdsContextData.class)
```
remove context for vlaue migrations, not required for this test class

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DataObjectMigratorValueMigrationTest.java:
```diff
@@ -0,0 +1,247 @@
+import org.junit.Test;
+
+/**
+ * Tests for {@link DoStructureMigrator}, with focus on data object value migrations ({@link IDoValueMigrationHandler})
```
Add .

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/DataObjectMigratorValueMigrationTest.java:
```diff
@@ -0,0 +1,247 @@
+        .withAreaInSquareMeter(10);
+
+    // no value migrations will match the provided data object
+    // RoomSizeValueMigrationHandler is ignored by default in s_migrationContext
```
RoomSizeValueMigrationHandler -> RoomSizeFixtureDoValueMigrationHandler_2

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/fixture/house/HouseFixtureDoValueMigrationHandler_1.java:
```diff
@@ -0,0 +1,61 @@
+import org.eclipse.scout.rt.platform.util.CollectionUtility;
+
+/**
+ * Replaces existing Room object, which contains values to be migrated
```
Add .
Check all classes.

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/fixture/house/HouseFixtureDoValueMigrationHandler_1.java:
```diff
@@ -0,0 +1,61 @@
+    RoomFixtureDo newRoom = BEANS.get(RoomFixtureDo.class)
+        .withName("migrated tiny room")
+        // should normally reference a constant such as RoomTypesFixture.ROOM to make sure it is a valid ID
+        // we intentionally insert a value which needs to be migrated by RoomTypeValueMigrationHandler
```
Fix name of RoomTypeValueMigrationHandler (add version)

stephan-merkli commented on org.eclipse.scout.rt.dataobject.test/src/test/java/org/eclipse/scout/rt/dataobject/migration/fixture/house/RoomTypesFixture.java:
```diff
@@ -0,0 +1,18 @@
+ */
+package org.eclipse.scout.rt.dataobject.migration.fixture.house;
+
+public final class RoomTypesFixture {
```
Use an Interface, for HouseType Fixtures too (no I prefix though).

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationInventory.java:
```diff
@@ -304,6 +329,44 @@ protected void validateDataObjectTypeVersion(IDoStructureMigrationHandler migrat
+   * @param valueMigrationHandler
+   *          Value migration handler to be validated
+   */
+  protected void validateValueMigrationHandler(Map<DoValueMigrationId, IDoValueMigrationHandler<?>> validatedValueMigrationHandlers, IDoValueMigrationHandler valueMigrationHandler) {
```
Switch argument order

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrationInventory.java:
```diff
@@ -304,6 +329,44 @@ protected void validateDataObjectTypeVersion(IDoStructureMigrationHandler migrat
+   *         handlers with the same type version are ordered in arbitrary, but deterministic order.
+   */
+  public List<IDoValueMigrationHandler<?>> getValueMigrationHandlers() {
+    return Collections.unmodifiableList(new ArrayList<>(m_valueMigrationHandlers.values()));
```
Maybe directly m_valueMigrationsHandlers.values().stream.collect(Collectors.toUnmodifiableList());?

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoStructureMigrator.java:
```diff
@@ -123,41 +119,80 @@ public boolean migrateDataObject(DoStructureMigrationContext ctx, IDataObject da
+    T typedDataObject = dataObjectMapper.readValue(migratedJson, valueType);
+
+    // Apply value migration
+    // Value migrations are applied _before_ data model migrations and may operate on data model related data objects.
```
remove comment about data model migrations

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoValueMigrationDataObjectVisitor.java:
```diff
@@ -0,0 +1,80 @@
+    m_ctx = ctx;
+
+    DoValueMigrationIdsContextData contextData = m_ctx.getGlobal(DoValueMigrationIdsContextData.class);
+    assertNotNull(contextData.getAppliedValueMigrationIds(), "Applied value migration IDs on context required.");
```
Assign to local variable?

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/DoValueMigrationId.java:
```diff
@@ -0,0 +1,36 @@
+    return new DoValueMigrationId(id);
+  }
+
+  private DoValueMigrationId(String id) {
```
Keep constructor with UUID, but convert within of method

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoValueMigrationHandler.java:
```diff
@@ -0,0 +1,151 @@
+ * <p>
+ * To avoid non-idempotent value migrations, do not re-use renamed or deleted ID values in the (future) code base. If
+ * non-idempotent value migrations cannot be avoided, consider using an {@link IDoStructureMigrationHandler} instead (if
+ * you control the namespace of the relevant data objects and are able to introduce a new type version).
```
you control -> you are in control

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoValueMigrationHandler.java:
```diff
@@ -0,0 +1,151 @@
+
+  /**
+   * @return Class of value node that needs to be migrated (assignable check is performed so ExampleId will also be
+   *         migrated in an IId node).
```
use {@link} for IId

stephan-merkli commented on org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/migration/IDoValueMigrationHandler.java:
```diff
@@ -0,0 +1,151 @@
+  /**
+   * Type version used for sorting.
+   * <p>
+   * The type version defined on a value migration is not related in any way to the type versions of the migrated data
```
Rephrase. They are not releated in a way that there are executed in between the structure migrations, but still related so that the same type versions are used for ordering.
Remove first sentence, use the second sentence as new first one. Then add "The value migrations are applied after all structure migrations, thus the type version doesn't directly related to a structure migration handler with the same type version."

stephan-merkli commented:
Jenkins please run tests

stephan-merkli commented:
Jenkins, please run tests again

LukasHuser committed:
Data object migration: Value migrations

New interface IDoValueMigrationHandler to support data object
value migrations. In contrast to IDoStructureMigrationHandler,
value migration handlers must not change the structure of a data object,
but can replace values of specific attributes.

A value migration handler operates on typed values. Values can be
of any type, typical use cases include merges, replacements or renaming
of built-in IDs, such as IUuIds. However, values can also be
IDoEntity data objects.

Value migration handlers are always applied after all
structure migration handlers have been applied. Value migrations are
applied in a specific order, defined by a type version.
Migration handlers with the same type version are applied according to
the order provided by the bean manager.

315763

stephan-merkli commented:
Jenkins please run tests
